### PR TITLE
feat : store user settings on server

### PR DESCRIPTION
**Pull request by amitiwary999**
Fixes issue #766 
 
store user setting on the server and made changes in setting according to new MVP design.
Fix back arrow not working in setting activity. 
Fix: Hotword detection does not start even on even when it is checked in the setting. 
Show them value name according to the retrieved theme from the server.
Add custom dialog to change server in case of anonymous mode.

### CODE REVIEW

**Comment by amitiwary999** [comment]
@chiragw15 @mayank408 review. Also I tried to solve setting problem as much as possible still if there will be some issue in setting, I will solve that  later.

File: app/src/main/java/org/fossasia/susi/ai/settings/contract/ISettingsPresenter.kt [hunk 1]
```diff
@@ -10,18 +8,22 @@ import org.fossasia.susi.ai.settings.ChatSettingsFragment
...
-    fun onAttach(chatSettingsFragment: ChatSettingsFragment)
-
     fun deleteMsg()
 
     fun getThemes(): String?
 
     fun setTheme(string: String)
 
-    fun enableMic(): Boolean
+    fun enablemic(): Boolean
```
Comment (mayank408) [inline app/src/main/java/org/fossasia/susi/ai/settings/contract/ISettingsPresenter.kt]: What the use of changing the function name?

File: app/src/main/java/org/fossasia/susi/ai/settings/ChatSettingsFragment.kt [hunk 2]
```diff
@@ -77,19 +81,40 @@ class ChatSettingsFragment : PreferenceFragmentCompat() {
...
         })
 
-        micSettings?.isEnabled = settingsPresenter?.enableMic() as Boolean
+        micSettings?.isEnabled = settingsPresenter.enablemic()
+
+        hotwordSettings?.isEnabled = settingsPresenter.enableHotword()
 
-        hotwordSettings?.isEnabled = settingsPresenter?.enableHotword() as Boolean
+        micSettings?.setOnPreferenceClickListener {
+            settingsPresenter.sendSetting("mic_input", (PrefManager.getBoolean(Constant.MIC_INPUT, false)).toString())
```
Comment (mayank408) [inline app/src/main/java/org/fossasia/susi/ai/settings/ChatSettingsFragment.kt]: Use Constants here not strings.

File: app/src/main/java/org/fossasia/susi/ai/rest/responses/susi/Settings.kt [hunk 3]
```diff
@@ -0,0 +1,56 @@
...
+ * Created by meeera on 30/6/17.
+ */
+
+class Settings {
+
+    /**
+     * Gets speechAlways
+     *
+     * @return the speechAlways
+     */
```
Comment (mayank408) [inline app/src/main/java/org/fossasia/susi/ai/rest/responses/susi/Settings.kt]: Avoid comments here, these are specific to java not kotlin. We are not using any getters and setters here.

File: app/src/main/java/org/fossasia/susi/ai/login/LoginPresenter.kt [hunk 4]
```diff
@@ -125,6 +127,28 @@ class LoginPresenter(loginActivity: LoginActivity): ILoginPresenter, ILoginModel
         }
     }
 
+    override fun onSuccessSetting(response: Response<UserSetting>) {
+
+        loginView?.showProgress(false)
```
Comment (amitiwary999) [inline app/src/main/java/org/fossasia/susi/ai/login/LoginPresenter.kt]: in that case, user logs in before retrieving setting and app will not show retrieved setting until user close and open app again.

Comment (chiragw15) [inline app/src/main/java/org/fossasia/susi/ai/login/LoginPresenter.kt]: No, in that case user won't be logged in but the progress dialog will dismiss. Nevermind, leave it. I thought something else. It's fine. Sorry, for trouble.

File: app/src/main/java/org/fossasia/susi/ai/settings/ChatSettingsFragment.kt (hunk 2 above)
Comment (amitiwary999) [inline app/src/main/java/org/fossasia/susi/ai/settings/ChatSettingsFragment.kt]: Which constant? on server we are saving setting value in string format. So I am converting the value of mic setting to string.

Comment (chiragw15) [inline app/src/main/java/org/fossasia/susi/ai/settings/ChatSettingsFragment.kt]: He means convert `"mic_input"` to Constant.MIC_INPUT

**Comment by amitiwary999** [comment]
@chiragw15 @mayank408  please review. I need it to make some important changes in setting.

File: app/src/main/java/org/fossasia/susi/ai/helper/PrefManager.java [hunk 5]
```diff
@@ -51,7 +51,7 @@ public static int getInt(String preferenceKey, int preferenceDefaultValue) {
         return getPreferences().getInt(preferenceKey, preferenceDefaultValue);
     }
 
-    public static void putTheme(String preferenceKey, String value){
+    public static void putTheme(String value){
```
Comment (mayank408) [inline app/src/main/java/org/fossasia/susi/ai/helper/PrefManager.java]: why you changed this?

Comment (amitiwary999) [inline app/src/main/java/org/fossasia/susi/ai/helper/PrefManager.java]: In putstring you already added key. Constant.Theme. so no need of providing key in parameter. It is specifically for theme.

File: app/src/main/java/org/fossasia/susi/ai/settings/ChatSettingsFragment.kt [hunk 6]
```diff
@@ -77,19 +81,40 @@ class ChatSettingsFragment : PreferenceFragmentCompat() {
...
+
+        speechAlways?.setOnPreferenceChangeListener { _, newValue ->
+            settingsPresenter.sendSetting(Constant.SPEECH_ALWAYS, newValue.toString())
+            true
+        }
+
+        speechOutput?.setOnPreferenceChangeListener { _, newValue ->
+            settingsPresenter.sendSetting(Constant.SPEECH_OUTPUT, newValue.toString())
+            true
+        }
```
Comment (chiragw15) [inline app/src/main/java/org/fossasia/susi/ai/settings/ChatSettingsFragment.kt]: There was listener for hotword detection too before shifting settings to MVP. Make sure everything which was there earlier, exists now too.

File: app/src/main/java/org/fossasia/susi/ai/rest/services/SusiService.java [hunk 7]
```diff
@@ -84,4 +86,26 @@
...
+                                                      @Query("access_token") String accessToken);
+
+    /**
+     * Get User settings
+     *
+     * @param accessToken
+     * @return settings
+     */
+    @GET("/aaa/listUserSettings.json")
+    Call<UserSetting> getUserSetting(@Query("access_token") String accessToken);
```
Comment (amitiwary999) [inline app/src/main/java/org/fossasia/susi/ai/rest/services/SusiService.java]: If I removed token then It is not working.

Comment (amitiwary999) [inline app/src/main/java/org/fossasia/susi/ai/rest/services/SusiService.java]: end point api.susi.ai/aaa/listUserSettings and api.susi.ai/aaa/listUserSettings?access_token both work. I am using second one. no problem I will change it.

File: app/src/main/java/org/fossasia/susi/ai/settings/ChatSettingsFragment.kt (hunk 6 above)
Comment (amitiwary999) [inline app/src/main/java/org/fossasia/susi/ai/settings/ChatSettingsFragment.kt]: there is a listener for checking hotword detection. Are you talking about storing it on the server?

Comment (amitiwary999) [inline app/src/main/java/org/fossasia/susi/ai/settings/ChatSettingsFragment.kt]: there is a listener for hotword detection. are you talking about storing it on server?

File: app/src/main/java/org/fossasia/susi/ai/rest/services/SusiService.java (hunk 7 above)
Comment (chiragw15) [inline app/src/main/java/org/fossasia/susi/ai/rest/services/SusiService.java]: No No. See whenever you are logged in, every call you make to susi is attached with access_token by interceptor. So, you don't need to explicitly provide it again as a parameter.

**Comment by amitiwary999** [comment]
I updated code.

File: app/src/main/java/org/fossasia/susi/ai/settings/ChatSettingsFragment.kt (hunk 6 above)
Comment (chiragw15) [inline app/src/main/java/org/fossasia/susi/ai/settings/ChatSettingsFragment.kt]: See this https://github.com/fossasia/susi_android/blob/c57f11fcfd3fe7f935ff11fd29c0da7a3be0589b/app/src/main/java/org/fossasia/susi/ai/activities/SettingsActivity.java#L211 and https://github.com/fossasia/susi_android/blob/c57f11fcfd3fe7f935ff11fd29c0da7a3be0589b/app/src/main/java/org/fossasia/susi/ai/activities/SettingsActivity.java#L223 .

Comment (chiragw15) [inline app/src/main/java/org/fossasia/susi/ai/settings/ChatSettingsFragment.kt]: Match current version with this https://github.com/fossasia/susi_android/blob/c57f11fcfd3fe7f935ff11fd29c0da7a3be0589b/app/src/main/java/org/fossasia/susi/ai/activities/SettingsActivity.java and see if everything is present and working fine.

**Comment by amitiwary999** [comment]
Yes every setting is added.

File: app/src/main/java/org/fossasia/susi/ai/rest/responses/susi/Settings.kt [hunk 8]
```diff
@@ -0,0 +1,32 @@
...
+
+/**
+ * Created by meeera on 30/6/17.
+ */
+
+class Settings {
+
+    @SerializedName("speech_always")
+    @Expose
+    val speechAlways: String ?= null
```
Comment (chiragw15) [inline app/src/main/java/org/fossasia/susi/ai/rest/responses/susi/Settings.kt]: This should be `Boolean`

File: app/src/main/java/org/fossasia/susi/ai/rest/responses/susi/Settings.kt [hunk 9]
```diff
@@ -0,0 +1,32 @@
...
+
+    @SerializedName("enter_send")
+    @Expose
+    val enterSend: String ?= null
```
Comment (chiragw15) [inline app/src/main/java/org/fossasia/susi/ai/rest/responses/susi/Settings.kt]: This too

File: app/src/main/java/org/fossasia/susi/ai/rest/responses/susi/Settings.kt [hunk 10]
```diff
@@ -0,0 +1,32 @@
...
+
+    @SerializedName("speech_output")
+    @Expose
+    val speechOutput: String ?= null
```
Comment (chiragw15) [inline app/src/main/java/org/fossasia/susi/ai/rest/responses/susi/Settings.kt]: This too

File: app/src/main/java/org/fossasia/susi/ai/rest/responses/susi/Settings.kt [hunk 11]
```diff
@@ -0,0 +1,32 @@
...
+
+    @SerializedName("mic_input")
+    @Expose
+    val micInput: String ?= null
```
Comment (chiragw15) [inline app/src/main/java/org/fossasia/susi/ai/rest/responses/susi/Settings.kt]: This too

File: app/src/main/java/org/fossasia/susi/ai/settings/ChatSettingsFragment.kt [hunk 12]
```diff
@@ -20,34 +21,36 @@ import org.fossasia.susi.ai.settings.contract.ISettingsPresenter
...
+    lateinit var settingsPresenter: ISettingsPresenter
 
     var textToSpeech: Preference? = null
     var rate: Preference? = null
     var server: Preference? = null
     var micSettings: Preference? = null
     var theme: ListPreference? = null
     var hotwordSettings: Preference? = null
-    var settingActivity: SettingsActivity? = null
     var utilModel: UtilModel?= null
```
Comment (chiragw15) [inline app/src/main/java/org/fossasia/susi/ai/settings/ChatSettingsFragment.kt]: unused

**Comment by chiragw15** [comment]
@amitiwary999 In [this](https://github.com/fossasia/susi_android/blob/c57f11fcfd3fe7f935ff11fd29c0da7a3be0589b/app/src/main/java/org/fossasia/susi/ai/activities/SettingsActivity.java) file , line number 211 to 240 checks if mobile supports mic input and hotword detection. If not, it disables their settings. I don't see code for that now. Please add that too

**Comment by amitiwary999** [comment]
@chiragw15 check line 89 and 91 in chatsetting fragment. mic and hotword are included.

**Comment by chiragw15** [comment]
Resolve conflicts. Rest is fine.

**Pull request merged**